File: DoublyCircular.cpp
#include <iostream>
using namespace std;

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node *PNODE;

class Doubly_CLL
{
private:
    PNODE Head;
    PNODE Tail;
    int iCount;

public:
    Doubly_CLL();
    void InsertFirst(int);
    void InsertLast(int);
    void DeleteFirst();
    void DeleteLast();
    void InsertAtPos(int, int);
    void DeleteAtPos(int);
    void Dispaly();
    int Count();
};

Doubly_CLL::Doubly_CLL()
{
    Head = NULL;
    Tail = NULL;
    iCount = 0;
}
void Doubly_CLL::InsertFirst(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;
    newn->prev = NULL;

    if ((Head == NULL) && (Tail == NULL))
    {
        Head = newn;
        Tail = newn;
    }
    else
    {
        newn->next = Head;
        Head->prev = newn;
        Head = newn;
    }
    Head->prev = Tail;
    Tail->next = Head;
}
void Doubly_CLL::InsertLast(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE;

    newn->data = iNo;
    newn->next = NULL;
    newn->prev = NULL;

    if ((Head == NULL) && (Tail == NULL))
    {
        Head = newn;
        Tail = newn;
    }
    else
    {
        Tail->next = newn;
        newn->prev = Tail;
        Tail = newn;
    }
    Head->prev = Tail;
    Tail->next = Head;
}
void Doubly_CLL::DeleteFirst()
{
    if ((Head == NULL) && (Tail == NULL))
    {
        return;
    }
    else if (Head == Tail)
    {
        delete Head;
        Head = NULL;
        Tail = NULL;
    }
    else
    {
        Head = Head->next;
        free(Tail->next);
    }
    Tail->next = Head;
    Head->prev = Tail;
}
void Doubly_CLL::DeleteLast()
{
    if ((Head == NULL) && (Tail == NULL))
    {
        return;
    }
    else if (Head == Tail)
    {
        delete Tail;
        Head = NULL;
        Tail = NULL;
    }
    else
    {
        Tail = Tail->prev;
        delete Tail->next;
    }
    Head->prev = Tail;
    Tail->next = Head;
}
void Doubly_CLL::InsertAtPos(int iNo, int pos)
{
    int iSize = 0;
    iSize = Count();
    if ((pos < 1) || (pos > iSize + 1))
    {
        return;
    }
    else if (pos == 1)
    {
        InsertFirst(iNo);
    }
    else if (pos == iSize + 1)
    {
        InsertLast(iNo);
    }
    else
    {
        int iCnt = 0;
        PNODE temp = Head;

        PNODE newn = NULL;
        newn = new NODE;

        newn->data = iNo;
        newn->next = NULL;
        newn->prev = NULL;

        for (iCnt = 1; iCnt < pos - 1; iCnt++)
        {
            temp = temp->next;
        }
        temp->next->prev = newn;
        newn->next = temp->next;
        newn->prev = temp;
        temp->next = newn;
    }
}
void Doubly_CLL::DeleteAtPos(int pos)
{
    int iSize = 0;
    iSize = Count();
    if ((pos < 1) || (pos > iSize))
    {
        return;
    }
    else if (pos == 1)
    {
        DeleteFirst();
    }
    else if (pos == iSize)
    {
        DeleteLast();
    }
    else
    {
        int iCnt = 0;
        PNODE temp = Head;

        for (iCnt = 1; iCnt < pos - 1; iCnt++)
        {
            temp = temp->next;
        }
        temp->next = temp->next->next;
        delete temp->next->prev;
        temp->next->prev = temp;
    }
}
void Doubly_CLL::Dispaly()
{
    PNODE temp = Head;
    cout << "Link list element is:" << endl;
    do
    {
        cout << "|" << temp->data << "|->";
        temp = temp->next;
    } while (temp != Head);
    cout << endl;
}
int Doubly_CLL::Count()
{
    int iCnt = 0;
    PNODE temp = Head;
    do
    {
        iCnt++;
        temp = temp->next;
    } while (temp != Head);

    return iCnt;
}

int main()
{
    int iRet = 0;
    Doubly_CLL obj;
    obj.InsertFirst(101);
    obj.InsertFirst(51);
    obj.InsertFirst(21);
    obj.InsertFirst(11);
    obj.Dispaly();

    obj.InsertLast(121);
    obj.Dispaly();

    obj.DeleteFirst();
    obj.DeleteLast();
    obj.Dispaly();

    obj.InsertAtPos(75, 3);
    obj.Dispaly();
    iRet = obj.Count();
    cout << "Number of node :" << iRet << endl;

    obj.DeleteAtPos(3);
    obj.Dispaly();
    iRet = obj.Count();
    cout << "Number of node :" << iRet << endl;

    return 0;
}

File: DoublyCircular.java
import java.lang.*;
import java.util.*;

class DoublyCircular
{
    public static void main(String arg[])
    {
        int iRet =0;
        DoublyCLL obj = new DoublyCLL();

        
        obj.InsertFirst(111);
        obj.InsertFirst(51);
        obj.InsertFirst(21);
        obj.InsertFirst(11);
        obj.Display();

        obj.InsertLast(121);
        obj.Display();

        obj.DeleteFirst();
        obj.DeleteLast();
        obj.Display();
        iRet= obj.Count();
        System.out.println("Nuber of node is:"+iRet);

        obj.InsertAtPos(75,3);
        obj.Display();
        iRet= obj.Count();
        System.out.println("Nuber of node is:"+iRet);

        obj.DeleteAtPos(3);
        obj.Display();
        iRet= obj.Count();
        System.out.println("Nuber of node is:"+iRet);


    }
}

class node 
{
    public int data;
    public node next;
    public node prev;
}

class DoublyCLL
{
    private node Head;
    private node Tail;

    public DoublyCLL()
    {
        Head = null;
        Tail = null;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node();

        newn.data = iNo;
        newn.next = null;
        newn.prev = null;

        if((Head==null)&&(Tail==null))
        {
            Head = newn;
            Tail = newn;
        }
        else
        {
           newn.next = Head;
            Head.prev = newn;
            Head = newn;
        }
        Head.prev = Tail;
        Tail.next = Head;
    }

    public void InsertLast(int iNo)
    {
        node newn = new node();

        newn.data = iNo;
        newn.next = null;

        if((Head==null)&&(Tail==null))
        {
            Tail = newn;
            Head = newn;
        }
        else
        {
            Tail.next = newn;
            newn.prev = Tail;
            Tail = newn;
        }
        Tail.next = Head;
        Head.prev = Tail;
    }

    public void DeleteFirst()
    {
        if((Head == null)||(Tail == null))
        {
            return;
        }
        else if(Head == Tail)
        {
            Head =null;
            Tail =null;
        }
        else
        {
            Head = Head.next;
            Tail.next=null;
        }
        Tail.next = Head;
        Head.prev = Tail;
    }

    public void DeleteLast()
    {
        if((Head == null)||(Tail == null))
        {
            return;
        } 
        else if(Head == Tail)
        {
            Head =null;
            Tail =null;
        }
        else
        {
           Tail = Tail.prev;
           Tail.next = null;
        }
        Head.prev = Tail;
        Tail.next = Head;
    }

    public void InsertAtPos(int iNo, int iPos)
    {
        int iSize = Count();
        if((iPos < 1)||(iPos > iSize+1))
        {
            System.out.println("Invalid Position");
            return;
        }
        else if(iPos == 1)
        {
            InsertFirst(iNo);
        }
        else if (iPos == iSize+1)
        {
            InsertLast(iNo);
        }
        else
        {
            int iCnt =0;
            node temp = Head;

            node newn = null;
            newn = new node();

            newn.data = iNo;
            newn.next = null;
            newn.prev = null;

            for(iCnt = 1; iCnt < iPos -1; iCnt++)
            {
                temp = temp.next;
            }
            temp.next.prev = newn;
            newn.next = temp.next; 
            newn.prev = temp;
            temp.next =newn;
        }

    }

    public void DeleteAtPos( int iPos)
    {
        int iSize = Count();
        if((iPos < 1)||(iPos > iSize))
        {
            System.out.println("Invalid Position");
            return;
        }
        else if(iPos == 1)
        {
            DeleteFirst();
        }
        else if (iPos == iSize)
        {
            DeleteLast();
        }
        else
        {
            int iCnt =0;
            node temp = Head;

            for(iCnt = 1; iCnt < iPos -1; iCnt++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;
            temp.next.prev = null;
            temp.next.prev = temp;
        }

    }

    public void Display()
    {
        node temp = Head;

        System.out.println("Link list element is:");
        do
        {
            System.out.print("|"+temp.data+"|->");
            temp = temp.next;
        } while(temp!=Head);
        
        System.out.println("");

    }

    public int Count()
    {
        int iCnt=0;
        node temp = Head;
        do
        {
           iCnt++;
            temp = temp.next;
        } while(temp!=Head);
        
        return iCnt;

    }
}